akka-http2-support/src/main/scala/akka/http/impl/engine/http2/parsing/HttpRequestHeaderHpackDecompression.scala

```diff
@@ -52,28 +52,26 @@ private[http2] final class HttpRequestHeaderHpackDecompression extends GraphStag
         // no backpressure (limited by SETTINGS_MAX_CONCURRENT_STREAMS)
         pull(streamIn)
 
-        def hasEndStream(frame: StreamFrameEvent): Boolean = frame match {
-          case h: HeadersFrame ⇒ h.endStream
-          case _               ⇒ false
-        }
-
-        // FIXME: ensure that this works for all combinations of events
-        //   (e.g. there can be an initial frame with endStream = true, endHeaders = false
-        //    in which case we need to wait for a continuation, etc.)
-        if (hasEndStream(httpSubStream.initialFrame)) {
-          val pushedRequest = processFrame(httpSubStream, httpSubStream.initialFrame)
+        httpSubStream.initialFrame match {
+          case h: HeadersFrame ⇒
+            if (h.endStream && !h.endHeaders) {
+              handleHeadersFromExpectedContinuations(httpSubStream, h)
+            } else if (h.endStream /* && h.endHeaders, implied by above if */ ) {
+              processFrame(httpSubStream, httpSubStream.initialFrame)
 
-          // we know frames should be empty here.
-          // but for sanity lets kill that stream anyway I guess (at least for now)
-          killFramesStream(httpSubStream)
-        } else {
-          processFrame(httpSubStream, httpSubStream.initialFrame)
-          // more data in following frames
+              // we know frames should be empty here.
+              // but for sanity lets kill that stream anyway I guess (at least for now)
+              killFramesStream(httpSubStream)
```

Comment:
Maybe we can replace it by a assertion instead? Or, we could already put this assertion into Http2SubStream?